cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/query/FlowQueries.kt

```diff
@@ -252,67 +326,90 @@ fun dataFlowWithValidator(
+                    contextSensitive = ContextSensitive in sensitivities,
+                )
+                .flatten()
+                .toSet()
+        val earlyTerminationPredicate = { n: Node, ctx: Context ->
+            earlyTermination?.let { it(n) } == true ||
+                (!allowOverwritingValue &&
+                    // TODO: This should be replaced with some check if the memory location/whatever
+                    // where the data is kept is (partially) written to.
+                    nodeToTrack in n.prevDFG &&
+                    (n as? Reference)?.access == AccessValues.WRITE)
+        }
+        val nextEOGEvaluation =
+            nodeToTrack.followEOGEdgesUntilHit(
+                collectFailedPaths = true,
+                findAllPossiblePaths = true,
+                scope = scope,
+                sensitivities = sensitivities,
+                earlyTermination = earlyTerminationPredicate,
+            ) {
+                predicate(it) && it in nextDFGPaths
+            }
+        allChildren +=
+            nextEOGEvaluation.failed.map { (failureReason, path) ->
+                SinglePathResult(
+                    value = false,
+                    children = mutableListOf(QueryTree(value = path)),
                     stringRepresentation =
-                        "The EOG path reached the node ${it.lastOrNull()} matching the required predicate" +
+                        "The EOG path reached the end  " +
                             if (earlyTermination != null)
-                                " before reaching a node matching the early termination predicate"
-                            else "",
-                    node = this,
+                                "(or ${path.lastOrNull()} which a predicate marking the end) "
+                            else
+                                "" +
+                                    "before passing through a node matching the required predicate.",
+                    node = nodeToTrack,
+                    terminationReason =
+                        if (failureReason == FailureReason.PATH_ENDED) {
+                            PathEnded(nodes.last())
+                        } else if (failureReason == FailureReason.HIT_EARLY_TERMINATION) {
+                            HitEarlyTermination(nodes.last())
+                        } else {
+                            StepsExceeded(nodes.last())
+                        },
                 )
-            }
+            } +
+                nextEOGEvaluation.fulfilled.map {
+                    SinglePathResult(
+                        value = true,
+                        children = mutableListOf(QueryTree(value = it)),
+                        stringRepresentation =
+                            "The EOG path reached the node ${it.lastOrNull()} matching the required predicate" +
+                                if (earlyTermination != null)
```

Comment:
Should we let this be fulfilled if early termination is != null or also considered it failed. If we keep it like this it is a general approximation, if we let it fail if early termination triggered, it would be a more conservative approach

Reply:
What? It only prints more text in the explanation. The check will pass/fail independently from this line of code

Reply:
Ah ok, I must have missjudged the line there and thought that It would not fail.